### PR TITLE
Fix DHCP v4/v6 race condition and memory leak

### PR DESCRIPTION
**Pull request by tony-josi-aws**
Description
-----------
This PR fixes DHCP - v4 and v6 race condition and memory leak.
While the DHCP state machine is processing response from a DHCP server, causing it to change the DHCP state, if another DHCP server responds to the DUT, state mismatch will cause the `vDHCPProcess` function to result in an infinite loop.

Thanks [Raghav](https://forums.freertos.org/u/Raghav) for reporting this issue in the [forum post](https://forums.freertos.org/t/code-stuck-in-the-vdhcpprocess-function-mostly-when-dhcp-state-machine-is-ewaitingacknowledge/22484).

Along with this fix, the zero copy buffers were not freed in the DHCPv6 state machine as well as in DHCPv4 corner cases, those are also corrected in this PR.

Test Steps
-----------

STM32F4 - DHCP v4 tests

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c [hunk 1]
```diff
@@ -341,6 +341,12 @@ void test_vDHCPProcess_ResetAndIncorrectStateWithSocketAlreadyCreated( void )
 
         /* Expect these arguments. */
         FreeRTOS_recvfrom_ExpectAndReturn( xDHCPv4Socket, NULL, 0UL, FREERTOS_ZERO_COPY + FREERTOS_MSG_PEEK, NULL, NULL, 0 );
+
+        /* Ignore the source address and source address
+         * length argument though. */
+        FreeRTOS_recvfrom_IgnoreArg_pxSourceAddress();
```
Comment (ActoryOu) [inline test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:347]: Is there any test case verifying multiple DHCP server scenarios?

Comment (tony-josi-aws) [reply test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:347]: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1226/files#diff-33d34c64f4cfe318de92341b204dd23da6d4594abd3a5a81d3a4f360be3a42f8R1347-R1437

**Review by ActoryOu — APPROVED** [review]

**Review by kar-rahul-aws — APPROVED** [review]
